ocaml/xapi/repository_helpers.ml

```diff
@@ -1478,3 +1526,83 @@ let set_pending_guidances ~ops ~coming =
            to_be_removed ;
          do_with_vm_pending_guidances ~op:ops.vm_add ~vm:vm_ref_str to_be_added
      )
+
+let failure_of_livepatch_component = function
+  | Livepatch.Xen ->
+      Guidance.RebootHostOnXenLivePatchFailure
+  | Livepatch.Kernel ->
+      Guidance.RebootHostOnKernelLivePatchFailure
+
+let component_of_livepatch_failure = function
+  | Guidance.RebootHostOnXenLivePatchFailure ->
+      Some Livepatch.Xen
+  | Guidance.RebootHostOnKernelLivePatchFailure ->
+      Some Livepatch.Kernel
+  | _ ->
+      None
+
+let merge_livepatch_failures ~previous_failures ~applied ~failed =
+  let current_failures = List.map failure_of_livepatch_component failed in
+  (* Determine how to deal with previous livepatch failures *)
+  List.fold_left
+    (fun acc_fails prev_fail ->
+      match component_of_livepatch_failure prev_fail with
+      | Some prev_failed_component -> (
+          let comp_str = Livepatch.string_of_component prev_failed_component in
+          match
+            ( List.mem prev_fail current_failures
+            , List.mem prev_failed_component applied
+            )
+          with
+          | true, false ->
+              debug "%s Livepatch failed again." comp_str ;
+              acc_fails
+          | false, true ->
+              (* applied in this update. No changes on current failures *)
+              debug
+                "%s Livepatch (failed in previous updates) has been applied."
+                comp_str ;
+              acc_fails
+          | true, true ->
+              (* Impossible case: a component is in both failed and applied *)
+              warn "%s Livepatch shouldn't be in both applied and failed lists."
+                comp_str ;
+              acc_fails
+          | false, false ->
+              (* Didn't try in this update. Keep previous failure in current list *)
+              debug "%s Livepatch failed in previous updates." comp_str ;
+              prev_fail :: acc_fails
+        )
+      | None ->
+          warn "Unknown livepatch failure." ;
+          acc_fails
+    )
+    current_failures previous_failures
+  |> fun failures ->
+  let open GuidanceSet in
+  let to_be_removed =
+    diff (of_list previous_failures) (of_list failures) |> elements
+  in
+  let to_be_added =
+    diff (of_list failures) (of_list previous_failures) |> elements
+  in
+  (to_be_removed, to_be_added)
+
+let update_livepatch_failures ~__context ~host ~applied ~failed =
+  let previous_failures =
+    Db.Host.get_pending_guidances_recommended ~__context ~self:host
+    |> List.map (fun g -> Guidance.of_pending_guidance g)
+    |> List.filter is_livepatch_failure
+  in
+  let to_be_removed, to_be_added =
+    merge_livepatch_failures ~previous_failures ~applied ~failed
+  in
```

Comment:
Can we simplify the logic of `merge_livepatch_failures`, I think the livepatch failure guidance `to_be_removed` is the guidance from `previous_failures` but got `applied`, and `to_be_added` can simply be the guidance in `failed` but not in `applied`(`Db.Host.add_pending_guidances_recommended` can endure element be added again as it is a Set).

Reply:
I prefer using the pattern match to get the whole thing be clear. And this function doesn't depend on any external logic like the set-like behavior of `Db.Host.add_pending_guidances_recommended`.